File: scripts/crowbar-testbuild.py
#!/usr/bin/python

# Copyright (c) 2015 SUSE LINUX GmbH, Nuernberg, Germany.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import argparse
import functools
import os
import re
import sys
import tempfile

import sh
from sh import Command

try:
    import jenkins
except ImportError:
    print("Please install 'python-jenkins' from pypi")
    sys.exit(1)


IBS_MAPPING = {
    'release/stoney/master': 'Devel:Cloud:4:Staging',
    'release/tex/master':    'Devel:Cloud:5:Staging',
    'stable/3.0':            'Devel:Cloud:6:Staging',
    'master':                'Devel:Cloud:7:Staging'
}

REPO_MAPPING = {
    'release/stoney/master': 'SLE_11_SP3',
    'release/tex/master':    'SLE_11_SP3',
    'stable/3.0':            'SLE_12_SP1',
    'master':                'SLE_12_SP2'
}

CLOUDSRC = {
    'release/stoney/master': 'develcloud4',
    'release/tex/master':    'develcloud5',
    'stable/3.0':            'develcloud6',
    'master':                'develcloud7'
}

MKCLOUD_HA_PARAMETERS = {
    'nodenumber': '4',
    'hacloud' :'1',
    'networkingmode': 'vxlan',
    'clusterconfig': '"data+services+network=2"'
}

MKCLOUD_HYPERV_PARAMETERS = {
    'networkingplugin': 'linuxbridge',
    'libvirt_type': 'hyperv',
    'networkingmode' :'vlan'
}

MKCLOUD_CEPH_PARAMETERS = {
    'nodenumber': '4',
    'want_ceph': '1',
    'networkingplugin': 'linuxbridge'
    }

JOB_PARAMETERS = {
    'crowbar-ha': MKCLOUD_HA_PARAMETERS,
    'crowbar-hyperv': MKCLOUD_HYPERV_PARAMETERS,
    'crowbar-ceph': MKCLOUD_CEPH_PARAMETERS,
    'barclamp-ceph': MKCLOUD_CEPH_PARAMETERS,
    'barclamp-hyperv': MKCLOUD_HYPERV_PARAMETERS,
    'barclamp-pacemaker': MKCLOUD_HA_PARAMETERS
}

htdocs_dir = '/srv/mkcloud'
htdocs_url = 'http://clouddata.cloud.suse.de/mkcloud/'

iosc = functools.partial(
    Command('/usr/bin/osc'), '-A', 'https://api.suse.de')


def ghs_set_status(repo, head_sha1, status):
    ghs = Command(
        os.path.abspath(
            os.path.join(os.path.dirname(sys.argv[0]),
                         'github-status/github-status.rb')))

    ghs('-r', 'crowbar/' + repo,
        '-c', head_sha1, '-a', 'set-status', '-s', status)

def _jenkins_read_api_credentials(cred_path):
    """file format for credentials is from jenkins-job-trigger perl script:
"""
    data = {}
    with open(cred_path, "r") as f:
        for line in f:
            m = re.match(r'\$(.*)="(.*)";', line)
            if m:
                key, val = m.group(1, 2)
                data[key] = val
    if 'SCHEME' not in data.keys():
        data['SCHEME'] = 'https'
    return data

def jenkins_job_trigger(repo, github_opts, cloudsource, ptfdir):
    print("triggering jenkins job with " + htdocs_url + ptfdir)
    creds = _jenkins_read_api_credentials('/etc/jenkinsapi.cred')
    server = jenkins.Jenkins('%s://%s' % (creds['SCHEME'], creds['JURL']),
                             username=creds['USER'], password=creds['APIKEY'])

    jenkins = Command(
        os.path.abspath(
            os.path.join(os.path.dirname(sys.argv[0]),
                         'jenkins/jenkins-job-trigger')))

    job_parameters = {
        'nodenumber': '2',
        'networkingplugin': 'openvswitch'
    }

    if repo in JOB_PARAMETERS:
        job_parameters = JOB_PARAMETERS[repo]

    job_parameters['cloudsource'] = cloudsource
    job_parameters['UPDATEREPOS'] = "%s%s" % (htdocs_url, ptfdir)
    job_parameters['github_pr'] = "crowbar/%s:%s" % (repo, github_opts)
    job_parameters['mkcloudtarget'] = "all_noreboot"

    server.build_job('openstack-mkcloud', **job_parameters)

def add_pr_to_checkout(repo, pr_id, head_sha1, pr_branch, spec):
    sh.curl(
        '-s', '-k', '-L',
        "https://github.com/crowbar/%s/compare/%s...%s.patch" % (
            repo, pr_branch, head_sha1),
        '-o', 'prtest.patch')
    sh.sed('-i', '-e', 's,Url:.*,%define _default_patch_fuzz 2,',
           '-e', 's,%patch[0-36-9].*,,', spec)
    Command('/usr/lib/build/spec_add_patch')(spec, 'prtest.patch')
    iosc('vc', '-m', "added PR test patch from %s#%s (%s)" % (
        repo, pr_id, head_sha1))


def prep_webroot(ptfdir):
    webroot = os.path.join(htdocs_dir, ptfdir)
    sh.rm('-rf', webroot)
    sh.mkdir('-p', webroot)
    return webroot


def trigger_testbuild(repo, github_opts):
    pr_id, head_sha1, pr_branch = github_opts.split(':')

    olddir = os.getcwd()
    ptfdir = repo + ':' + github_opts
    webroot = prep_webroot(ptfdir)
    workdir = tempfile.mkdtemp()
    build_failed = False

    try:
        pkg = repo if repo.startswith("crowbar") else "crowbar-" + repo
        spec = pkg + '.spec'

        try:
            os.chdir(workdir)
            buildroot = os.path.join(os.getcwd(), 'BUILD')
            iosc('co', IBS_MAPPING[pr_branch], pkg, '-c')
            os.chdir(pkg)
            add_pr_to_checkout(repo, pr_id, head_sha1, pr_branch, spec)
            iosc('build', '--root', buildroot, '--noverify', '--noservice',
                 REPO_MAPPING[pr_branch], 'x86_64', spec, _out=sys.stdout)
        except:
            build_failed = True
            print("Build failed: " + str(sys.exc_info()[1]))
            raise
        else:
            sh.cp('-p', sh.glob(
                os.path.join(buildroot, '.build.packages/RPMS/*/*.rpm')),
                webroot)
        finally:
            os.chdir(olddir)
            sh.cp('-p', os.path.join(buildroot, '.build.log'),
                  os.path.join(webroot, 'build.log'))
    finally:
        sh.sudo.rm('-rf', workdir)

    if not build_failed:
        jenkins_job_trigger(
            repo, github_opts, CLOUDSRC[pr_branch], ptfdir)

    ghs_set_status(
        repo, head_sha1,
        'failure' if build_failed else'pending')


def main():
    parser = argparse.ArgumentParser(
        description='Build a testpackage for a crowbar/ Pull Request')
    parser.add_argument('repo', metavar='REPO',
                        help='github repo in the crowbar organization')
    parser.add_argument('pr', metavar='PRID:SHA1:BRANCH',
                        help='github PR id, SHA1 head of PR, and '
                             'destination branch of PR')

    args = parser.parse_args()

    trigger_testbuild(args.repo, args.pr)


if __name__ == '__main__':
    main()

File: scripts/crowbar-testbuild.rb
#!/usr/bin/env ruby

# Copyright (c) 2017 SUSE LINUX GmbH, Nuernberg, Germany.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

require "yaml"
require "optparse"
require "fileutils"
require "tmpdir"

class CrowbarTestbuild
  def initialize(options, config)
    @config = config
    @options = options
    @repo = @options[:repo] || ""
    @org = @options[:org] || ""
    @pr = @options[:pr] || ""
  end

  def parameters(flavor="")
    p = @config["parameters_template"]["default"]
    p.merge(@config["parameters_template"][flavor] || {})
  end

  def pr_split
    @pr.split(":")
  end

  def pr_id
    pr_split[0]
  end

  def pr_sha1
    pr_split[1]
  end

  def pr_sha1_short
    pr_split[1][0,8]
  end

  def pr_branch
    pr_split[2]
  end

  def job_name
    "#{@org}/#{@repo} PR #{pr_id} #{pr_sha1_short}"
  end

  def mkcloud_target
    @config["mkcloud"]["target"]
  end

  def mkcloud_job_name
    @config["mkcloud"]["job_name"]
  end

  def branch_map
    @config["branches"][pr_branch]
  end

  def bs_project
    branch_map["bs_project"]
  end

  def bs_repo
    branch_map["bs_repo"]
  end

  def cloudsource
    branch_map["cloudsource"]
  end

  def pr_string
    [@org, @repo, @pr].join(":")
  end

  def ptf_dir
    unex = File.join(@config["infrastructure"]["htdocs_dir"], pr_string)
    File.expand_path(unex)
  end

  def ptf_url
    @config["infrastructure"]["htdocs_url"] + pr_string + "/"
  end

  def extra_build_parameters(mode)
    p = @config["parameters_template"]["default"]
    p.merge!(@config["parameters"]["#{@org}/#{@repo}"][mode]) if @config["parameters"].include?("#{@org}/#{@repo}")
    p
  end

  def trigger_jenkins_job(mode)
    j = @config["infrastructure"]["jenkins_job_trigger"]
    sub_context = mode == "standard" ? "":mode
    sub_context_suffix = sub_context.empty? ? "" : ":#{sub_context}"
    job_name = "#{@org}/#{@repo} PR #{pr_id} #{pr_sha1_short} #{sub_context}"
    cmd = j["cmd"]
    cmd = File.join(File.dirname(__FILE__), cmd) unless File.exists?(cmd)
    jcmd = [ cmd, mkcloud_job_name, "-p" ]
    para_default = {
      mkcloudtarget: mkcloud_target,
      cloudsource: cloudsource,
      github_pr: "#{@org}/#{@repo}:#{@pr}#{sub_context_suffix}",
      UPDATEREPOS: ptf_url,
      job_name: job_name
    }
    para_extra = extra_build_parameters(mode)
    p = para_default.merge(para_extra).map { |k,v| "#{k}=#{v}" }
    jcmd += p
    puts "Triggering jenkins job with url #{ptf_url} and directory #{ptf_dir}"
    system(*jcmd) or raise
  end

  def jenkins_jobs_scenarios
    if @config["parameters"].include?("#{@org}/#{@repo}")
      @config["parameters"]["#{@org}/#{@repo}"].keys
    else
      @config["parameters"]["default"].keys
    end
  end

  def trigger_jenkins_jobs
    jenkins_jobs_scenarios.each do |k|
      trigger_jenkins_job(k)
    end
  end

  def osc_cmd(subcommand)
    [ @config["infrastructure"]["osc"]["cmd"] ] +
      ( @config["infrastructure"]["osc"]["parameters"]["_global"] || [] ) +
      [ subcommand ] +
      ( @config["infrastructure"]["osc"]["parameters"][subcommand] || [] )
  end

  def package_name
    pack = @repo
    pack = "crowbar-" + pack unless pack.start_with?("crowbar")
    pack
  end

  def package_spec_file
    package_name + ".spec"
  end

  def add_pr_to_checkout(compare_method)
    puts "Start crowbar testbuild."
    puts "Github diff method: #{compare_method}"
    patch_diff = "prtest.#{compare_method}"
    patch_diff_url = "https://github.com/#{@org}/#{@repo}/compare/#{pr_branch}...#{pr_sha1}.#{compare_method}"
    puts "Fetching patch from this URL: #{patch_diff_url}"
    system(
      *[
        "curl", "-s", "-k", "-L", "-o", patch_diff, patch_diff_url
      ]
    )

    system(
      *[
        "sed", "-i",
        "-e", "s,Url:.*,%define _default_patch_fuzz 2,",
        "-e", "s,%patch[0-36-9].*,,", package_spec_file
      ]
    )

    system(
      *[
        "/usr/lib/build/spec_add_patch", package_spec_file, patch_diff
      ]
    )

    system(
      *(
        osc_cmd("vc") +
        [
          "-m",
          "added PR test patch from #{@org}/#{@repo}##{pr_id} (#{pr_sha1})"
        ]
      )
     )
  end

  def build(method)
    FileUtils.rm_rf(ptf_dir)
    FileUtils.mkdir_p(ptf_dir)
    work_dir = Dir.mktmpdir
    build_root = "BUILD"

    Dir.chdir(work_dir) do
      begin
        osc_co = osc_cmd("co") + [bs_project, package_name]
        system(*osc_co)
        Dir.chdir(package_name) do
          Dir.mkdir(build_root)
          add_pr_to_checkout(method)
          system(
            *(
              osc_cmd("build") +
              [
                "--root", File.join(Dir.pwd, build_root),
                '--noverify', '--noservice',
                bs_repo, "x86_64", package_spec_file
              ]
            )
          ) or raise SystemCallError, "Build failed."
        end
      rescue => e
        puts e.message
        puts e.backtrace
        return false
      else
        # copy rpms
        FileUtils.cp(
          Dir.glob(File.join(package_name, build_root, ".build.packages/RPMS/*/*.rpm")),
          ptf_dir, :preserve => true
        ) or raise
        return true
      ensure
        # copy log
        FileUtils.cp(
          File.join(package_name, build_root, ".build.log"),
          File.join(ptf_dir, "build.log")
        )
        FileUtils.rm_rf(work_dir)
      end
    end
  end

end

conf_dirs = [ "/etc/mkcloud", Dir.pwd, File.dirname(__FILE__) ]
conf_file_name = "crowbar-testbuild.yaml"
## options
options = {}
optparse = OptionParser.new do |opts|
  opts.banner = "Build a testpackage for a crowbar pull request"
  opts.on("-o", "--org GITHUB_ORG", "Github Organisation") do |org|
    options[:org] = org
  end
  opts.on("-r", "--repo GITHUB_REPO", "Github Repository") do |repo|
    options[:repo] = repo
  end
  opts.on("-p", "--pr PR_STRING", "Github PR string='PRID:SHA1:DESTINATION_BRANCH'") do |pr|
    options[:pr] = pr
  end
  opts.on("-t", "--trigger-gating", "Trigger gating job run with jenkins-job-trigger") do
    options[:trigger_gating] = true
  end
  opts.on("-c", "--config CONFIG_FILE", "Config file or directory to search for file: #{conf_file_name}") do |c|
    conf_dirs = [c]
  end
end
optparse.parse!

## config
begin
  c = conf_dirs.shift
  c = File.join(c, conf_file_name) if File.directory?(c)
  puts "Trying config file: #{c}"
  config = YAML.load_file(c)
rescue Psych::SyntaxError
  puts "Error: Invalid YAML syntax in config file: #{c}"
  exit 2
rescue
  retry if conf_dirs.size > 0
  puts "Error: Could not find a config file: #{conf_file_name}"
  exit 3
end

## main
ctb = CrowbarTestbuild.new(options, config)
methods = [ "diff", "patch" ]
begin
  m = methods.shift
  puts "Building with method #{m}."
  ctb.build(m) or raise "Buidling with method #{m} failed."
  ctb.trigger_jenkins_jobs or raise "Could not trigger gating job" if options[:trigger_gating]
rescue => e
  puts e.message
  puts e.backtrace
  retry if methods.size > 0
  exit 1
else
  puts "Building with method #{m} successful."
end
